### ISSUE TITLE
FR: Make coverbrowser's listmenu available for plugin authors

### ISSUE THREAD

**Issue by Billiam** [koreader/koreader#12793]
**Does your feature request involve difficulty completing a task? Please describe.**

Plugins that want to display a list of books are likely to use the `menu` widget currently. However, this widget supports a simple text string for the title, and single info value on the right. The list menu by comparison:

* Supports images
* Both title and subtitle/authors
* Multiple info fields.

For displaying books, whether from the file system, or remote data, this format is a huge improvement, and would be of use to plugin authors.

**Describe the solution you'd like**

I would like to see the listmenu provided as a generic widget (without BookInfoManager), that allowed a user to provide a list of records, possibly with lazy loading for images that could be managed outside of the widget.

**Describe alternatives you've considered**

For my own project, I've vendored the listmenu (and covermenu) for use in displaying remote book data (used to map the current document to a published book), and then removed some calls to BookInfoWidget and replaced those with values from the passed in items.

This "works" but is fragile, since underlying dependencies can change and cause failures.

**Comment by Frenzie** [comment]
It's probably not super fragile (put another way, it hasn't changed all that much over the years), but agreed if someone has the time that some parts of CoverBrowser would preferably move to core.

**Comment by Billiam** [comment]
> It's probably not super fragile (put another way, it hasn't changed all that much over the years)

I can't say how often it has, or would be an issue, just my immediate experience was in updating from 2024.07 to 2024.11

The `header_padding` value was removed from `menu.lua` in 9cd305177e223853b9af1fe1ba4e325e2520a24e, which the 2024.07 `listmenu` used in `_recalculateDimen`, which caused a failure when using the previously working plugin in the newer KOReader. The fix was to update the vendored code to the `2024.11` versions as well, but the effect is that:

* Either additional dependencies that may change (like menu.lua) _also_ need to be vendored, even if no changes need to be made there, or
* Vendored files need to be (much more extensively) modified to maintain backwards compatibility in unedited areas, (I don't think this is feasible in the long term), or
* Accept that this approach is fragile, and a plugin release is going to be KOReader version-specific, and can easily fail in future updates.

**Comment by poire-z** [comment]
> Accept that this approach is fragile, and a plugin release is going to be KOReader version-specific, and can easily fail in future updates.

I'm afraid you and you users will have to live with that :/
It's already painful to track what other parts need to be changed when we make a change (ie. header_padding, supposed to fix some geometry consistency issues) - without having to think about external plugins.
I also would not like us to have to provide a stable API/behaviour - and get stuck to it (or documenting it), without even knowing about plugins in the wild using it or not.

For the original listmenu.lua request : it's currently part of the coverbrowser plugin, so normally not usable by anything else.
If you need it, with possible tweaks: just duplicate it. Even: try to make a generic widget with the behaviour you want (and the behaviour CoverBrowser needs, but listmenu just hacks the existing menu.lua - for the context it is used by FileChooser, so it's not really standalone). May be it will be good enough to be used for something else, and if by a little bit of core (ie. OPDS, calibre, stuff I don't really knwow about), it could be shipped with core.
Again, I'd rather not have options/parameters not used by core, but possibly by external plugins, that we will never test and never think about when adjusting the behaviour/features that we do use.

Allowing external user plugins is a side effect :) there's no garantee, it shouldn't be with strings attached for us :)

**Comment by NiLuJe** [comment]
Having something ListMenu-like in core does sound tempting, but it would take a potentially significant amount of work, as the current implementing is fairly heavily tied to CoverBrowser (for better *and* for worse ;)).

It's also fairly old, hasn't seen much action over the years, and is based on possibly the oldest and most complex base widget we have, so it's, err, not for the faint of heart.

TL;DR: It might make more sense to start from scratch than try to mangle ListMenu into submission. (As opposed to an ideal world where there would be a glorious refactor with a new-ish public widget, and CoverBrowser adapted to use that, scrapping its own Listmenu horror).
